File: app/src/main/java/me/ycdev/android/devtools/apps/selector/AppsSelectorAdapter.java
package me.ycdev.android.devtools.apps.selector;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.view.View;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;

import me.ycdev.android.devtools.R;
import me.ycdev.android.lib.common.apps.AppInfo;
import me.ycdev.android.lib.commonui.base.ListAdapterBase;
import me.ycdev.android.lib.commonui.base.ViewHolderBase;

class AppsSelectorAdapter extends ListAdapterBase<AppInfo, AppsSelectorAdapter.ViewHolder> {
    public interface SelectedAppsChangeListener {
        void onSelectedAppsChanged(int newCount);
    }

    private SelectedAppsChangeListener mChangeListener;
    private boolean mMultiChoice;
    private HashSet<AppInfo> mSelectedApps = new HashSet<>();

    private View.OnClickListener mCheckedChangeListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            AppInfo item = (AppInfo) v.getTag();
            item.isSelected = !item.isSelected;
            if (item.isSelected) {
                if (!mMultiChoice && mSelectedApps.size() > 0) {
                    getOneSelectedApp().isSelected = false;
                    mSelectedApps.clear();
                }
                mSelectedApps.add(item);
            } else {
                mSelectedApps.remove(item);
            }
            notifyDataSetChanged();
            mChangeListener.onSelectedAppsChanged(mSelectedApps.size());
        }
    };

    public AppsSelectorAdapter(Context cxt, SelectedAppsChangeListener listener,
                               boolean multiChoice) {
        super(cxt);
        mChangeListener = listener;
        mMultiChoice = multiChoice;
    }

    public int getSelectedAppsCount() {
        return mSelectedApps.size();
    }

    public List<AppInfo> getSelectedApps() {
        return new ArrayList<>(mSelectedApps);
    }

    public AppInfo getOneSelectedApp() {
        Iterator<AppInfo> it = mSelectedApps.iterator();
        if (it.hasNext()) {
            return it.next();
        }
        return null;
    }

    @Override
    public void setData(List<AppInfo> data) {
        super.setData(data);
        Collections.sort(mList, new AppInfo.AppNameComparator());
    }

    @Override
    protected @LayoutRes int getItemLayoutResId() {
        return R.layout.apps_selector_list_item;
    }

    @NonNull
    @Override
    protected ViewHolder createViewHolder(@NonNull View itemView, int position) {
        return new ViewHolder(itemView, position);
    }

    @Override
    protected void bindView(@NonNull AppInfo item, @NonNull ViewHolder vh) {
        vh.iconView.setImageDrawable(item.appIcon);
        vh.appNameView.setText(item.appName);
        vh.pkgNameView.setText(item.pkgName);
        vh.checkBox.setChecked(item.isSelected);
        vh.checkBox.setTag(item);
        vh.checkBox.setOnClickListener(mCheckedChangeListener);
    }

    static class ViewHolder extends ViewHolderBase {
        public ImageView iconView;
        public TextView appNameView;
        public TextView pkgNameView;
        public CheckBox checkBox;

        public ViewHolder(View itemView, int position) {
            super(itemView, position);
            iconView = (ImageView) itemView.findViewById(R.id.app_icon);
            appNameView = (TextView) itemView.findViewById(R.id.app_name);
            pkgNameView = (TextView) itemView.findViewById(R.id.pkg_name);
            checkBox = (CheckBox) itemView.findViewById(R.id.checkbox);
        }
    }
}

File: app/src/main/java/me/ycdev/android/devtools/apps/selector/AppsSelectorAdapter.kt
package me.ycdev.android.devtools.apps.selector

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import me.ycdev.android.devtools.R
import me.ycdev.android.devtools.apps.selector.AppsSelectorAdapter.ViewHolder
import me.ycdev.android.devtools.databinding.AppsSelectorListItemBinding
import me.ycdev.android.lib.common.apps.AppInfo
import java.util.Collections
import java.util.Comparator

open class AppsSelectorAdapter(
    private val changeListener: SelectedAppsChangeListener,
    multiChoice: Boolean,
) : RecyclerView.Adapter<ViewHolder>() {
    interface SelectedAppsChangeListener {
        fun onSelectedAppsChanged(newCount: Int)
    }

    private val checkedChangeListener =
        OnClickListener { v ->
            val item = v.tag as AppInfo
            val changedPackageNames = selectionState.toggle(item.pkgName)
            changedPackageNames.forEach { pkgName ->
                val position = data?.indexOfFirst { it.pkgName == pkgName } ?: -1
                if (position >= 0) {
                    notifyItemChanged(position)
                }
            }
            changeListener.onSelectedAppsChanged(selectionState.selectedCount)
        }

    private val selectionState = AppSelectionState(multiChoice)
    val selectedAppsCount: Int
        get() = selectionState.selectedCount
    val selectedApps: List<AppInfo>
        get() =
            selectionState.selectedPackageNames.mapNotNull { pkgName ->
                data?.firstOrNull { it.pkgName == pkgName }
            }

    val oneSelectedApp: AppInfo?
        get() = selectionState.oneSelectedPackageName?.let { pkgName ->
            data?.firstOrNull { it.pkgName == pkgName }
        }

    var data: List<AppInfo>? = null

    fun sort(comparator: Comparator<AppInfo>) {
        data?.let {
            Collections.sort(it, comparator)
            notifyItemRangeChanged(0, it.size)
        }
    }

    private fun getItem(position: Int): AppInfo = data!![position]

    override fun getItemCount(): Int = data?.size ?: 0

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): ViewHolder {
        val itemView =
            LayoutInflater
                .from(parent.context)
                .inflate(R.layout.apps_selector_list_item, parent, false)
        return ViewHolder(itemView)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int,
    ) {
        val item = getItem(position)
        holder.binding.appIcon.setImageDrawable(item.appIcon)
        holder.binding.appName.text = item.appName
        holder.binding.pkgName.text = item.pkgName
        holder.binding.checkbox.isChecked = selectionState.isSelected(item.pkgName)
        holder.binding.checkbox.tag = item
        holder.binding.checkbox.setOnClickListener(checkedChangeListener)
    }

    class ViewHolder(
        itemView: View,
    ) : RecyclerView.ViewHolder(itemView) {
        val binding: AppsSelectorListItemBinding = AppsSelectorListItemBinding.bind(itemView)
    }
}
